src/goto-symex/complexity_module.h

```diff
@@ -0,0 +1,117 @@
+
+/// \file
+/// Symbolic Execution
+
+#ifndef CPROVER_GOTO_SYMEX_COMPLEXITY_MODULE_H
+#define CPROVER_GOTO_SYMEX_COMPLEXITY_MODULE_H
+
+#include "complexity_violation.h"
+#include "goto_symex_state.h"
+#include "symex_heuristic_transformation.h"
+
+/// Symex complexity module.
+///
+/// Dynamically cancels branches when symex considers them too complex.
+/// A branch is too complex when runtime may be beyond reasonable
+/// bounds due to internal structures becoming too large or solving conditions
+/// becoming too big for the SAT solver to deal with.
+///
+/// On top of the branch cancelling there is special consideration for loops.
+/// As branches get cancelled it keeps track of what loops are currently active
+/// up through the stack. If a loop has too many of its child branches killed
+/// (including additional loops or recursion) it won't do any more unwinds of
+/// that particular loop and blacklist it.
+///
+/// This blacklist is only held for as long as any loop is still active. As soon
+/// as the loop iteration ends and the context in which the code is being
+/// executed changes it'll be able to be run again.
+///
+/// Example of loop blacklisting:
+///
+///     loop A: (complexity: 5070)
+///         loop B: (complexity: 5000)
+///         loop C: (complexity: 70)
+///
+/// In the above loop B will be blacklisted if we have a complexity limitation
+/// < 5000, but loop A and B will still be run, because when loop B is removed
+/// the complexity of the loop as a whole is acceptable.
+class complexity_modulet
+{
+public:
+  complexity_modulet() = default;
+
+  explicit complexity_modulet(message_handlert &logger) : log(logger)
+  {
+  }
+
+  complexity_modulet(message_handlert &logger, const optionst &options);
+
+  /// Is the complexity module active?
+  /// \return Whether complexity limits are active or not.
+  inline bool complexity_limits_active()
+  {
+    return this->complexity_active;
+  }
+
+  /// Checks the passed-in state to see if its become too complex for us to deal
+  /// with, and if so set its guard to false.
+  /// \param state goto_symex_statet you want to check the complexity of.
+  /// \return True/false depending on whether this branch should be abandoned.
+  complexity_violationt check_complexity(goto_symex_statet &state);
+
+  /// Runs a suite of transformations on the state and symex executable,
+  /// performing whatever transformations are required depending on
+  /// the modules that have been loaded.
+  void run_transformations(
+    complexity_violationt heuristic_result,
+    goto_symex_statet &current_state);
+
+protected:
+  mutable messaget log;
+
+  /// Is the complexity module active, usually coincides with a max_complexity
+  /// value above 0.
+  bool complexity_active = false;
+
+  /// Functions called when the heuristic has been violated. Can perform
+  /// any form of branch, state or full-program transformation.
+  std::vector<symex_heuristic_transformationt> violation_transformations;
```

Comment:
`symex_heuristics_transformationt` -> `symex_complexity_limit_exceeded_actiont`. This isn't the heuristic, it's specifically the action to take when a limit (whether heuristic or not) has been broken